lambda/translquote.ml

```diff
@@ -0,0 +1,3144 @@
+      let lbl = quote_record_field env lid.loc lbl in
+      let exp = quote_expression transl stage exp in
+      Exp_desc.setfield loc rcd lbl exp
+    | Texp_array (_, _, exps, _) ->
+      let exps = List.map (quote_expression transl stage) exps in
+      Exp_desc.array loc exps
+    | Texp_ifthenelse (cond, then_, else_) ->
+      let cond = quote_expression transl stage cond in
+      let then_ = quote_expression transl stage then_ in
+      let else_ = Option.map (quote_expression transl stage) else_ in
+      Exp_desc.ifthenelse loc cond then_ else_
+    | Texp_sequence (exp1, _, exp2) ->
+      let exp1 = quote_expression transl stage exp1 in
+      let exp2 = quote_expression transl stage exp2 in
+      Exp_desc.sequence loc exp1 exp2
+    | Texp_while wh ->
+      let cond = quote_expression transl stage wh.wh_cond in
+      let body = quote_expression transl stage wh.wh_body in
+      Exp_desc.while_ loc cond body
+    | Texp_for floop ->
+      let low = quote_expression transl stage floop.for_from
+      and high = quote_expression transl stage floop.for_to
+      and dir =
+        match floop.for_dir with
+        | Asttypes.Upto -> true
+        | Asttypes.Downto -> false
+      and name = quote_name (mkloc (Ident.name floop.for_id) loc) in
+      with_new_idents_values [floop.for_id];
+      let body = quote_expression transl stage floop.for_body in
+      without_idents_values [floop.for_id];
+      Exp_desc.for_simple loc (quote_loc loc) name low high dir
+        (Lam.func Var_value extract floop.for_id body)
+    | Texp_send (obj, meth, _) ->
+      let obj = quote_expression transl stage obj in
+      let meth = quote_method loc meth in
+      Exp_desc.send loc obj meth
+    | Texp_open _ -> fatal_error "No support for opening modules yet."
+    | Texp_letmodule (ident, _, _, mod_exp, body) -> (
+      let mod_exp = quote_module_exp transl stage loc mod_exp in
+      match ident with
+      | None ->
+        Exp_desc.letmodule_nonbinding loc mod_exp
+          (quote_expression transl stage body)
+      | Some ident ->
+        let name = quote_name (mkloc (Ident.name ident) loc) in
+        with_new_idents_modules [ident];
+        let body = quote_expression transl stage body in
+        without_idents_modules [ident];
+        Exp_desc.letmodule loc (quote_loc loc) name mod_exp
+          (Lam.func Var_module extract ident body))
+    | Texp_assert (exp, _) ->
+      let exp = quote_expression transl stage exp in
+      Exp_desc.assert_ loc exp
+    | Texp_lazy exp ->
+      let exp = quote_expression transl stage exp in
+      Exp_desc.lazy_ loc exp
+    | Texp_quotation exp ->
+      let exp = quote_expression transl (stage + 1) exp in
+      Exp_desc.quote loc exp
+    | Texp_antiquotation exp ->
+      if stage > 0
+      then
+        let exp = quote_expression transl (stage - 1) exp in
+        Exp_desc.antiquote loc exp
+      else Exp_desc.splice loc (Code.inject (transl exp))
```

Comment:
I'm slightly worried there might be surprising orders of evaluations of splices arising from this, but I haven't checked if there actually are.

Reply:
Let's discuss offline.